src/UglyToad.PdfPig/Parser/CatalogFactory.cs

```diff
@@ -102,8 +104,16 @@ public void Increment()
 
             do
             {
-                var current = toProcess.Dequeue();
-
+                var current = toProcess.Dequeue();
+                var currentReferenceHash = current.reference.GetHashCode();                
+                if (visitedTokens.Contains(currentReferenceHash))

```

Comment:
:-) Have faith in the hashcode :-)  

Okay. Okay. If not could change **visitedTokens** from 

HashSet<int>   to   Dictionary<long, HashSet<int>> 

which stored each object number directly and in the result the list of generations directly.

First cut (and untest) at code it's less clear (if inline) and takes more cpu and memory. Might be able to make it read better with a bit more work.
`
               var currentReferenceObjectNumber = current.reference.ObjectNumber;
                var currentReferenceGeneration = current.reference.Generation;
                if (visitedTokens.ContainsKey(currentReferenceObjectNumber))
                {
                    var generations = visitedTokens[currentReferenceObjectNumber];
                   
                    if (generations.Contains(currentReferenceGeneration))
                    {
                        continue; // don't revisit token already processed. break infinite loop. Issue #512
                    }
                    else
                    {
                        generations.Add(currentReferenceGeneration);
                        visitedTokens[currentReferenceObjectNumber] = generations;
                    }
                }
                else
                {
                    visitedTokens.Add(currentReferenceObjectNumber, new HashSet<int>() { currentReferenceGeneration });

                    visitedTokensWorkingWindow.Enqueue((currentReferenceObjectNumber, currentReferenceGeneration));
                    if (visitedTokensWorkingWindow.Count==1000)
                    {
                        var toBeRemovedFromWorkingHashset = visitedTokensWorkingWindow.Dequeue();
                        var toBeRemovedObjectNumber = toBeRemovedFromWorkingHashset.ObjectNumber;
                        var toBeRemovedGeneration = toBeRemovedFromWorkingHashset.Generation;
                        var generations = visitedTokens[toBeRemovedObjectNumber];
                        generations.Remove(toBeRemovedGeneration);
                        if (generations.Count ==0)
                        {
                            visitedTokens.Remove(toBeRemovedObjectNumber); 
                        }
                        else
                        {
                            visitedTokens[toBeRemovedObjectNumber] = generations;
                        }
                    }
                }

`

As an optional extra could keep a Queue and so reduce the memory used by looking at only the last 1000 tokens however if a loop is larger than that it would be missed. Perhaps length of working window bould be parsing option.

` var visitedTokensWorkingWindow = new Queue<(long ObjectNumber, int Generation)>(1000);`